scene/3d/mesh_instance_3d.cpp

```diff
@@ -515,6 +515,108 @@ bool MeshInstance3D::_property_get_revert(const StringName &p_name, Variant &r_p
+
+	Ref<ArrayMesh> bake_mesh;
+
+	if (p_existing.is_valid()) {
+		ERR_FAIL_NULL_V_MSG(p_existing, Ref<ArrayMesh>(), "The existing mesh must be a valid ArrayMesh.");
+		ERR_FAIL_COND_V_MSG(source_mesh == p_existing, Ref<ArrayMesh>(), "The source mesh can not be the same mesh as the existing mesh.");
+
+		bake_mesh = p_existing;
+	} else {
+		bake_mesh.instantiate();
+	}
+
+	ERR_FAIL_NULL_V_MSG(skin_ref, Ref<ArrayMesh>(), "The source mesh must have a valid skin.");
+	ERR_FAIL_NULL_V_MSG(skin_internal, Ref<ArrayMesh>(), "The source mesh must have a valid skin.");
+	RID skeleton = skin_ref->get_skeleton();
+	ERR_FAIL_COND_V_MSG(!skeleton.is_valid(), Ref<ArrayMesh>(), "The source mesh must have its skin registered with a valid skeleton.");
+
+	const int bone_count = RenderingServer::get_singleton()->skeleton_get_bone_count(skeleton);
+	ERR_FAIL_COND_V(bone_count <= 0, Ref<ArrayMesh>());
+	ERR_FAIL_COND_V(bone_count < skin_internal->get_bind_count(), Ref<ArrayMesh>());
+
+	LocalVector<Transform3D> bone_transforms;
+	bone_transforms.resize(bone_count);
+	for (int bone_index = 0; bone_index < bone_count; bone_index++) {
+		bone_transforms[bone_index] = RenderingServer::get_singleton()->skeleton_bone_get_transform(skeleton, bone_index);
+	}
+
+	bake_mesh->clear_surfaces();
+
+	int mesh_surface_count = source_mesh->get_surface_count();
+
+	for (int surface_index = 0; surface_index < mesh_surface_count; surface_index++) {
+		ERR_CONTINUE(source_mesh->surface_get_primitive_type(surface_index) != Mesh::PRIMITIVE_TRIANGLES);
+
+		uint32_t surface_format = source_mesh->surface_get_format(surface_index);
+
+		ERR_CONTINUE(0 == (surface_format & Mesh::ARRAY_FORMAT_VERTEX));
+		ERR_CONTINUE(0 == (surface_format & Mesh::ARRAY_FORMAT_BONES));
+		ERR_CONTINUE(0 == (surface_format & Mesh::ARRAY_FORMAT_WEIGHTS));
+
+		unsigned int bones_per_vertex = surface_format & Mesh::ARRAY_FLAG_USE_8_BONE_WEIGHTS ? 8 : 4;
+
+		surface_format &= ~Mesh::ARRAY_FORMAT_BONES;
+		surface_format &= ~Mesh::ARRAY_FORMAT_WEIGHTS;
+
+		const Array &source_mesh_arrays = source_mesh->surface_get_arrays(surface_index);
+
+		ERR_FAIL_COND_V(source_mesh_arrays.size() != RS::ARRAY_MAX, Ref<ArrayMesh>());
+
+		const Vector<Vector3> &source_mesh_vertex_array = source_mesh_arrays[Mesh::ARRAY_VERTEX];
+		const Vector<int> &source_mesh_bones_array = source_mesh_arrays[Mesh::ARRAY_BONES];
+		const Vector<float> &source_mesh_weights_array = source_mesh_arrays[Mesh::ARRAY_WEIGHTS];
+
+		unsigned int vertex_count = source_mesh_vertex_array.size();
+		int expected_bone_array_size = vertex_count * bones_per_vertex;
+		ERR_CONTINUE(source_mesh_bones_array.size() != expected_bone_array_size);
+		ERR_CONTINUE(source_mesh_weights_array.size() != expected_bone_array_size);
+
+		Array new_mesh_arrays;
+		new_mesh_arrays.resize(Mesh::ARRAY_MAX);
+		for (int i = 0; i < source_mesh_arrays.size(); i++) {
+			if (i == Mesh::ARRAY_VERTEX || i == Mesh::ARRAY_BONES || i == Mesh::ARRAY_WEIGHTS) {
```

Comment:
Should the normals not be rotated with the bones?

Reply:
added